From 6d7f6ad49192fd570fd13a6c554681885efca291 Mon Sep 17 00:00:00 2001
From: Matthew <matthew@urgero.net>
Date: Fri, 15 Dec 2023 12:50:56 -0600
Subject: [PATCH 192/196] Add activation code for Joshua upgrade (#310)

Make all necessary preparations for the Joshua upgrade Dec 22, 2023.

Used commit 6e2e115 as a template for this commit.

---------
Co-authored-by: Shammah Chancellor <213740+schancel@users.noreply.github.com>
---
 src/chainparams.cpp                           |  8 ++++
 src/consensus/activation.cpp                  | 10 +++++
 src/consensus/activation.h                    |  4 ++
 src/consensus/addresses.h                     |  4 ++
 src/consensus/params.h                        |  4 ++
 src/init.cpp                                  |  3 +-
 src/minerfund.cpp                             |  5 +++
 src/validation.cpp                            |  2 +-
 .../logos_feature_minerfund_activation.py     | 42 ++++++++++++++++++-
 9 files changed, 78 insertions(+), 4 deletions(-)

diff --git a/src/chainparams.cpp b/src/chainparams.cpp
index f8ee3b9d8..ebcbdca29 100644
--- a/src/chainparams.cpp
+++ b/src/chainparams.cpp
@@ -113,6 +113,8 @@ public:
         consensus.deuteronomyActivationTime = 1687359480;
         // 2023-12-22T03:27:00.000Z protocol upgrade
         consensus.joshuaActivationTime = 1703215620;
+        // 2024-06-20T20:51:00.000Z protocol upgrade
+        consensus.judgesActivationTime = 1718916660;
 
         /**
          * The message start string is designed to be unlikely to occur in
@@ -256,6 +258,10 @@ public:
         consensus.joshuaActivationTime =
             mainnetConsensus.joshuaActivationTime -
             testnetActivationOffset;
+        // 2024-06-20T20:51:00.000Z protocol upgrade
+        consensus.judgesActivationTime =
+            mainnetConsensus.judgesActivationTime -
+            testnetActivationOffset;
 
         // "ltdk" with MSB set
         diskMagic[0] = 0xec;
@@ -368,6 +374,8 @@ public:
             mainnetConsensus.deuteronomyActivationTime;
         // 2023-12-22T03:27:00.000Z protocol upgrade
         consensus.joshuaActivationTime = mainnetConsensus.joshuaActivationTime;
+        // 2024-06-20T20:51:00.000Z protocol upgrade
+        consensus.judgesActivationTime = mainnetConsensus.judgesActivationTime;
 
         // "lrdk" with MSB set
         diskMagic[0] = 0xec;
diff --git a/src/consensus/activation.cpp b/src/consensus/activation.cpp
index fa501d269..9b0a4b777 100644
--- a/src/consensus/activation.cpp
+++ b/src/consensus/activation.cpp
@@ -52,3 +52,13 @@ bool IsDeuteronomyEnabled(const Consensus::Params &params,
     return pindexPrev->GetMedianTimePast() >=
            gArgs.GetArg("-deuteronomyactivationtime", params.deuteronomyActivationTime);
 }
+
+bool IsJoshuaEnabled(const Consensus::Params &params,
+                          const CBlockIndex *pindexPrev) {
+    if (pindexPrev == nullptr) {
+        return false;
+    }
+
+    return pindexPrev->GetMedianTimePast() >=
+           gArgs.GetArg("-joshuaactivationtime", params.joshuaActivationTime);
+}
diff --git a/src/consensus/activation.h b/src/consensus/activation.h
index 4cccc3036..16e2eee2b 100644
--- a/src/consensus/activation.h
+++ b/src/consensus/activation.h
@@ -31,4 +31,8 @@ bool IsNumbersEnabled(const Consensus::Params &params,
 bool IsDeuteronomyEnabled(const Consensus::Params &params,
                           const CBlockIndex *pindexPrev);
 
+/** Check if December 22nd, 2023 protocol upgrade has activated. */
+bool IsJoshuaEnabled(const Consensus::Params &params,
+                          const CBlockIndex *pindexPrev);
+
 #endif // BITCOIN_CONSENSUS_ACTIVATION_H
diff --git a/src/consensus/addresses.h b/src/consensus/addresses.h
index a265d305d..2753e1340 100644
--- a/src/consensus/addresses.h
+++ b/src/consensus/addresses.h
@@ -133,5 +133,9 @@ Consensus::CoinbaseAddresses AddressSets = {
             "lotus_16PSJHA7AxHkiV4XACtMRSzKmw8F1pT3ZbsEnzJ4s",
             "lotus_16PSJHA7AxHkiV4XACtMRSzKmw8F1pT3ZbsEnzJ4s",
         },
+    .joshua =
+        {
+            "lotus_16PSJMzCdB851yS4m59RUG5CKfjhJxTSWBo8EK1os",
+        },
 };
 } // namespace RewardAddresses
diff --git a/src/consensus/params.h b/src/consensus/params.h
index 6ee4c8fe6..9a6294b66 100644
--- a/src/consensus/params.h
+++ b/src/consensus/params.h
@@ -20,6 +20,7 @@ struct CoinbaseAddresses {
     std::vector<std::string> leviticus;
     std::vector<std::string> numbers;
     std::vector<std::string> deuteronomy;
+    std::vector<std::string> joshua;
 };
 
 /**
@@ -43,6 +44,9 @@ struct Params {
     /** Unix time used for MTP activation of 2023-12-22T03:27:00.000Z protocol
      * upgrade */
     int joshuaActivationTime;
+    /** Unix time used for MTP activation of 2024-06-20T20:51:00.000Z protocol
+     * upgrade */
+    int judgesActivationTime;
 
     /**
      * Don't warn about unknown BIP 9 activations below this height.
diff --git a/src/init.cpp b/src/init.cpp
index 4c2549ad6..ef3a6e912 100644
--- a/src/init.cpp
+++ b/src/init.cpp
@@ -425,7 +425,8 @@ void SetupServerArgs(NodeContext &node) {
         "-uiplatform",
         // TODO remove after the December 2021 upgrade
         "-exodusactivationtime", "-leviticusactivationtime",
-        "-numbersactivationtime", "-deuteronomyactivationtime"};
+        "-numbersactivationtime", "-deuteronomyactivationtime",
+        "-joshuaactivationtime"};
 
     // Set all of the args and their help
     // When adding new options to the categories, please keep and ensure
diff --git a/src/minerfund.cpp b/src/minerfund.cpp
index eab1aab40..25e3c83c9 100644
--- a/src/minerfund.cpp
+++ b/src/minerfund.cpp
@@ -56,6 +56,11 @@ std::vector<CTxOut> GetMinerFundRequiredOutputs(const Consensus::Params &params,
         return {};
     }
 
+    if (IsJoshuaEnabled(params, pindexPrev)) {
+        return BuildOutputsCycling(params.coinbasePayoutAddresses.joshua,
+                                   pindexPrev, blockReward);
+    }
+
     if (IsDeuteronomyEnabled(params, pindexPrev)) {
         return BuildOutputsCycling(params.coinbasePayoutAddresses.deuteronomy,
                                    pindexPrev, blockReward);
diff --git a/src/validation.cpp b/src/validation.cpp
index fb51c77f4..8b71d13c1 100644
--- a/src/validation.cpp
+++ b/src/validation.cpp
@@ -278,7 +278,7 @@ bool CheckSequenceLocks(const CTxMemPool &pool, const CTransaction &tx,
 static bool IsReplayProtectionEnabled(const Consensus::Params &params,
                                       int64_t nMedianTimePast) {
     return nMedianTimePast >= gArgs.GetArg("-replayprotectionactivationtime",
-                                           params.joshuaActivationTime);
+                                           params.judgesActivationTime);
 }
 
 static bool IsReplayProtectionEnabled(const Consensus::Params &params,
diff --git a/test/functional/logos_feature_minerfund_activation.py b/test/functional/logos_feature_minerfund_activation.py
index ff433f45a..4cb3eb1cb 100755
--- a/test/functional/logos_feature_minerfund_activation.py
+++ b/test/functional/logos_feature_minerfund_activation.py
@@ -30,8 +30,9 @@ LEVITICUS_ACTIVATION_TIME = 2010000000
 NUMBERS_ACTIVATION_TIME = 2020000000
 DEUTERONOMY_ACTIVATION_TIME = 2030000000
 JOSHUA_ACTIVATION_TIME = 2040000000
+JUDGES_ACTIVATION_TIME = 2050000000
 
-REPLAYPROTECTION_ACTIVATION_TIME = JOSHUA_ACTIVATION_TIME
+REPLAYPROTECTION_ACTIVATION_TIME = JUDGES_ACTIVATION_TIME
 
 # see consensus/addresses.h, use getaddressinfo to get the scriptPubKey
 GENESIS_SCRIPTS = [
@@ -118,7 +119,9 @@ DEUTERONOMY_SCRIPTS = [
     "76a914017cc7d3bb5512f21835cac23bd92b361393dcd288ac",
     "76a914017cc7d3bb5512f21835cac23bd92b361393dcd288ac",
 ]
-
+JOSHUA_SCRIPTS = [
+    "76a9149f3248570eff3658b794f29e6e59f394eca8f69988ac",
+]
 
 class MinerFundActivationTest(BitcoinTestFramework):
     def set_test_params(self):
@@ -130,6 +133,7 @@ class MinerFundActivationTest(BitcoinTestFramework):
             f'-leviticusactivationtime={LEVITICUS_ACTIVATION_TIME}',
             f'-numbersactivationtime={NUMBERS_ACTIVATION_TIME}',
             f'-deuteronomyactivationtime={DEUTERONOMY_ACTIVATION_TIME}',
+            f'-joshuaactivationtime={JOSHUA_ACTIVATION_TIME}',
             f'-replayprotectionactivationtime={REPLAYPROTECTION_ACTIVATION_TIME}',
         ]]
 
@@ -310,6 +314,40 @@ class MinerFundActivationTest(BitcoinTestFramework):
             prepare_block(block)
             assert_equal(node.submitblock(ToHex(block)), None)
 
+        # Using the joshua addresses before upgrade fails
+        block = make_block_cb_post_numbers(JOSHUA_SCRIPTS)
+        prepare_block(block)
+        assert_equal(node.submitblock(ToHex(block)), 'bad-cb-minerfund')
+        node.setmocktime(JOSHUA_ACTIVATION_TIME)
+
+        # Mine 11 blocks with JOSHUA_ACTIVATION in the middle
+        # That moves MTP exactly to JOSHUA_ACTIVATION
+        for i in range(-6, 6):
+            block = make_block_cb_post_numbers(DEUTERONOMY_SCRIPTS)
+            block.nTime = JOSHUA_ACTIVATION_TIME + i
+            prepare_block(block)
+            assert_equal(node.submitblock(ToHex(block)), None)
+
+        assert_equal(node.getblockchaininfo()['mediantime'],
+                     JOSHUA_ACTIVATION_TIME)
+        
+        # Now the using the genesis, exodus, leviticus, or numbers, and deuteronomy addresses fails
+        for block in [
+            make_block_with_cb_scripts(GENESIS_SCRIPTS),
+            make_block_with_cb_scripts(EXODUS_SCRIPTS),
+            make_block_with_cb_scripts(LEVITICUS_SCRIPTS),
+            make_block_cb_post_numbers(NUMBERS_SCRIPTS),
+            make_block_cb_post_numbers(DEUTERONOMY_SCRIPTS),
+        ]:
+            prepare_block(block)
+            assert_equal(node.submitblock(ToHex(block)), 'bad-cb-minerfund')
+       
+        # Using the Joshua scripts now works
+        for i in range(0, 12):
+            block = make_block_cb_post_numbers(JOSHUA_SCRIPTS)
+            prepare_block(block)
+            assert_equal(node.submitblock(ToHex(block)), None)
+
         # Check replay protection is not enabled yet
         tx = CTransaction()
         tx.vin = [CTxIn(COutPoint(int(cointxid, 16), 1))]
-- 
2.39.5 (Apple Git-154)

